Refactor the houston Package

AmFlint opened the pull request:
## Description

Refactor the `houston` package to be a high-level interface to interact with the Houston GraphQL API.

Clean the code, and make it more maintainable.

## 🎟 Issue(s)

https://github.com/astronomer/issues/issues/4161

## 🧪 Functional Testing

> List the functional testing steps to confirm this feature or fix.

## 📸 Screenshots

> Add screenshots to illustrate the validity of these changes.

## 📋 Checklist

- [x] Rebased from the main (or release if patching) branch (before testing)
- [x] Ran `make test` before taking out of draft
- [x] Added/updated applicable tests
- [x] Tested against [Houston-API](https://github.com/astronomer/houston-api/) and [Astronomer](https://github.com/astronomer/astronomer/) (if necessary).
- [ ] Communicated to/tagged owners of respective clients potentially impacted by these changes.
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

andriisoldatenko commented on auth/auth.go:
```diff
@@ -17,22 +17,12 @@ import (
 
 // basicAuth handles authentication with the houston api
-func basicAuth(username, password string) (string, error) {
+func basicAuth(username, password string, client houston.HoustonClientInterface) (string, error) {
```
```suggestion
func basicAuth(username, password string, client houston.ClientInterface) (string, error) {
```

wdyt to rename to ClientInterface? since it's already in houston package.

andriisoldatenko commented on houston/auth.go:
```diff
@@ -0,0 +1,29 @@
+		return nil, err
+	}
+	return acResp.Data.GetAuthConfig, nil
+}
```
```suggestion
}

```

andriisoldatenko commented on houston/houston.go:
```diff
@@ -17,12 +17,73 @@ var (
+}
+
+// HoustonClientImplementation - implementation of the Houston Client Interface
+type HoustonClientImplementation struct {
```
houston.Houston a bit wordy IMO.

andriisoldatenko replied:
```suggestion
type ClientImplementation struct {
```

andriisoldatenko commented on houston/houston.go:
```diff
@@ -17,12 +17,73 @@ var (
 )
 
+// HoustonClientInterface - Interface that defines methods exposed by the Houston API
+type HoustonClientInterface interface {
```
```suggestion
type ClientInterface interface {
```

andriisoldatenko commented on houston/serviceaccount.go:
```diff
@@ -0,0 +1,98 @@
+package houston
```
rename file to service_account.go pls

andriisoldatenko commented on houston/serviceaccount.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// CreateServiceAccountInWorkspace - create a service account into a workspace
+func (h HoustonClientImplementation) CreateServiceAccountInWorkspace(variables CreateServiceAccountRequest) (*WorkspaceServiceAccount, error) {
```
```suggestion
func (h HoustonClientImplementation) CreateWorkspaceServiceAccount(variables CreateServiceAccountRequest) (*WorkspaceServiceAccount, error) {
```

andriisoldatenko commented on houston/serviceaccount.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// DeleteServiceAccountFromDeployment - delete a service account from a workspace
+func (h HoustonClientImplementation) DeleteServiceAccountFromDeployment(deploymentID, serviceAccountID string) (*ServiceAccount, error) {
```
```suggestion
func (h HoustonClientImplementation) DeleteDeploymentServiceAccount(deploymentID, serviceAccountID string) (*ServiceAccount, error) {
```

andriisoldatenko commented:
@AmFlint pls resolve conflicts.

AmFlint commented on houston/houston.go:
```diff
@@ -18,12 +18,73 @@ var (
+
+// Init - initialized the Houston Client singleton object with proper HTTP Client configuration
+// set as a variable so we can change it to return mock houston clients
+// TODO: RENAME THIS METHOD ONCE TESTS ARE REFACTORED TOO
```
Sorry about that, I renamed the `Init` method to `NewClient`, and the old `NewHoustonClient` method to `newInternalClient` (as this becomes a structure internal to the houston package).

neel-astro commented on Makefile:
```diff
@@ -59,6 +59,10 @@ uninstall:
 
+# TODO: WE MIGHT WANT TO ADD A TOP-LEVEL mock COMMAND TO GENERATE ALL OF THE MOCK INTERFACES
+mock_houston:
+	mockery --filename=ClientInterface.go --output=airflow/mocks --dir=houston --name ClientInterface
```
sorry I didn't realise this earlier, but it would be better to keep mocks of an interface either where interface is defined or where it is being used, so you might want to move this mocks to `houston` folder instead of `airflow`

AmFlint replied:
Yeah I agree with you, I just wanted to keep the same conventions already configured in the project.

neel-astro reviewed (approved):
lgtm 👍 , have added a minor comment regarding placing mocks, also do have a look if there are any easy ways to bring up the dropping code coverage in this PR.

AmFlint commented:
The code coverage drops in this PR because I've added more files.

I added and refactored tests in the other issue.

However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response

andriisoldatenko commented:
> The code coverage drops in this PR because I've added more files.
> 
> I added and refactored tests in the other issue.
> 
> However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response

i don't understand you point, i think we want keep coverage as is. Please add tests to fix code coverage at least.

AmFlint commented:
> > The code coverage drops in this PR because I've added more files.
> > I added and refactored tests in the other issue.
> > However, I don't see a value in testing every Houston method since all it does is call doWithClient and return a structure from the response
> 
> i don't understand you point, i think we want keep coverage as is. Please add tests to fix code coverage at least.

I added tests for the houston package
